Account for sealed traits in privacy and trait bound errors

estebank opened the pull request:
On trait bound errors caused by super-traits, identify if the super-trait is publicly accessibly and if not, explain "sealed traits".

```
error[E0277]: the trait bound `S: Hidden` is not satisfied
  --> $DIR/sealed-trait-local.rs:17:20
   |
LL | impl a::Sealed for S {}
   |                    ^ the trait `Hidden` is not implemented for `S`
   |
note: required by a bound in `Sealed`
  --> $DIR/sealed-trait-local.rs:3:23
   |
LL |     pub trait Sealed: self::b::Hidden {
   |                       ^^^^^^^^^^^^^^^ required by this bound in `Sealed`
   = note: `Sealed` is a "sealed trait", because to implement it you also need to implelement `a::b::Hidden`, which is not accessible; this is usually done to force you to use one of the provided types that already implement it
```

Deduplicate privacy errors that point to the same path segment even if their deduplication span are different.

When encountering a path that is not reachable due to privacy constraints path segments other than the last, keep metadata for the last path segment's `Res` in order to look for alternative import paths for that item to suggest. If there are none, be explicit that the item is not accessible.

```
error[E0603]: module `b` is private
  --> $DIR/re-exported-trait.rs:11:9
   |
LL | impl a::b::Trait for S {}
   |         ^ private module
   |
note: the module `b` is defined here
  --> $DIR/re-exported-trait.rs:5:5
   |
LL |     mod b {
   |     ^^^^^
help: consider importing this trait through its public re-export instead
   |
LL | impl a::Trait for S {}
   |      ~~~~~~~~
```

```
error[E0603]: module `b` is private
  --> $DIR/private-trait.rs:8:9
   |
LL | impl a::b::Hidden for S {}
   |         ^  ------ trait `b` is not publicly reachable
   |         |
   |         private module
   |
note: the module `b` is defined here
  --> $DIR/private-trait.rs:2:5
   |
LL |     mod b {
   |     ^^^^^
```

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5284883544/jobs/9562795809) [(plain)](https://github.com/rust-lang/rust/commit/0d5859c5f545efbf856c534667fc55dceef11d92/checks/14307842052/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'rust-lang/simpleinfra@master' (SHA:5aa3313339663c490d58f9face5433f8c68a95e9)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180116 sha256=351235b2326fb4db7a18e257e13ce7896c5f77339521e2c2612e71e154800a19
  Stored in directory: /tmp/pip-ephem-wheel-cache-mplhx5xi/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:c2ffa5b151e8074fa1eb19a2cc293987595005e7fb19468585afaa69a4ee38b4
Uploading finished image to https://ci-caches.rust-lang.org/docker/e51156f19850ce886cec818e46dc2a021e0aa7c270d15673e8fe74cd8522fc8ac3995109aebb688ee49ed586735ed4cf5f8c06d44c48298fa09c35ae4b082281

<botocore.awsrequest.AWSRequest object at 0x7f1d812ae950>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_resolve/src/ident.rs:1627: \
Use a single space after dots in comments.
some tidy checks failed
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5284941964/jobs/9562920823) [(plain)](https://github.com/rust-lang/rust/commit/844092c2f2f8b9379a484a65b6207bbb7b09ee27/checks/14307998479/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'rust-lang/simpleinfra@master' (SHA:5aa3313339663c490d58f9face5433f8c68a95e9)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180116 sha256=351235b2326fb4db7a18e257e13ce7896c5f77339521e2c2612e71e154800a19
  Stored in directory: /tmp/pip-ephem-wheel-cache-5v_09mok/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:659239ce4e658b21e7c03e0f095b18260d71f1f42367f2e2462919af05900a74
Uploading finished image to https://ci-caches.rust-lang.org/docker/e51156f19850ce886cec818e46dc2a021e0aa7c270d15673e8fe74cd8522fc8ac3995109aebb688ee49ed586735ed4cf5f8c06d44c48298fa09c35ae4b082281

<botocore.awsrequest.AWSRequest object at 0x7fa4fbff47d0>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_resolve/src/ident.rs:1627: \
Use a single space after dots in comments.
some tidy checks failed
```

</details>

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2724,6 +2724,28 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                        };
+                        if let DefKind::Trait = tcx.def_kind(item_def_id) && !visible_item {
+                            err.note(format!(
+                                "`{short_item_name}` is a \"sealed trait\", because to `impl` it \
```
```suggestion
                                "`{short_item_name}` is a \"sealed trait\", because to implement it \
```

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2724,6 +2724,28 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                        if let DefKind::Trait = tcx.def_kind(item_def_id) && !visible_item {
+                            err.note(format!(
+                                "`{short_item_name}` is a \"sealed trait\", because to `impl` it \
+                                you also need to `impl {}`, which is not accessible; \
```
```suggestion
                                you also need to implement {}, which is not accessible; \
```

compiler-errors commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2724,6 +2724,28 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                                "`{short_item_name}` is a \"sealed trait\", because to `impl` it \
+                                you also need to `impl {}`, which is not accessible; \
+                                this is usually done to force you to use one of the provided \
+                                types that already `impl` it",
```
```suggestion
                                types that already implement it",
```

compiler-errors commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -737,13 +737,18 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     // For better failure detection, pretend that the import will
                     // not define any names while resolving its module path.
                     let orig_vis = import.vis.take();
+                    // let mut full_path = import.module_path.clone();
```
can you prune the commented code?

compiler-errors commented on tests/ui/macros/issue-88228.stderr:
```diff
@@ -4,7 +4,7 @@ error: cannot find macro `bla` in this scope
    |     ^^^
    |
-help: consider importing this macro
+help: consider importing this macro through the public re-export
```
```suggestion
help: consider importing this macro through its public re-export
```

compiler-errors commented on tests/ui/privacy/sealed-traits/private-trait-non-local.stderr:
```diff
@@ -0,0 +1,9 @@
+error[E0603]: module `index` is private
+  --> $DIR/private-trait-non-local.rs:2:18
+   |
+LL | use core::slice::index::private_slice_index::Sealed;
```
I quite like this framing, though I wonder if we should either point to somewhere with privacy rules, or at least drive this home by saying something like "an item is only reachable via a path if all of the segments are reachable". Maybe better wording, but something that folks can take away from the error message as intuition for why public-unreachable messages happen the way they do?

estebank replied:
I'd like to add some note about it, but I'm not sure what it should look like at the moment.

petrochenkov commented:
This touches name resolution infra, not only diagnostics.
r? @petrochenkov

petrochenkov commented:
You guys really want me to suffer.

In a typical estebank's fashion the PR
1. makes several independent changes but merges them together, so it's not clear a) what effect each change has on diagnostics (test outputs) and b) what cost benefit ratio each independent change has.
2. introduces ad hoc code in random places and technical debt so I have to understand which changes are not worth it and are better dropped, which I cannot do due to `1.`
3. has no PR description summarizing what you want to achieve

As a result, I don't want to see this PR merged due to `2.` but I also don't want to review it to suggest improvements due to `1-3.`.

petrochenkov commented:
Could you split this into parts small enough to make it digestable?

- one commit introducing new tests
- several commits introducing *minimal* orthogonal changes to name resolution infra (non-diagnostics.rs code, this is what I want to keep maintainable and garbage-free) and showing their effect on test outputs
- commit with remaining changes in diagnostics.rs files

(And also add a PR description.)
@rustbot author

estebank committed:
Account for sealed traits in trait bound errors

When implementing a public trait with a private super-trait, we now emit
a note that the missing bound is not going to be able to be satisfied,
and we explain the concept of a sealed trait.

estebank committed:
Tweak privacy errors to account for reachable items

Suggest publicly accessible paths for items in private mod:

  When encountering a path in non-import situations that are not reachable
  due to privacy constraints, search for any public re-exports that the
  user could use instead.

Track whether an import suggestion is offering a re-export.

When encountering a path with private segments, mention if the item at
the final path segment is not publicly accessible at all.

Add item visibility metadata to privacy errors from imports:

  On unreachable imports, record the item that was being imported in order
  to suggest publicly available re-exports or to be explicit that the item
  is not available publicly from any path.

  In order to allow this, we add a mode to `resolve_path` that will not
  add new privacy errors, nor return early if it encounters one. This way
  we can get the `Res` corresponding to the final item in the import,
  which is used in the privacy error machinery.

khionu commented:
@petrochenkov This review was really in bad form. You have valid critique that could have carried its own weight without either of the first two lines. Slipping in a public, passive-aggressive callout is no way to talk things out with your fellow project members. Please avoid that in the future, and if you feel you need help resolving issues with other people in the community, please reach out to Mods. We're here to help issues before they escalate, not just after.

estebank commented:
I've split this PR to the best of my abilities into easily digestible units. That being done, I do not believe initial PR was unreasonably unfocused, the changes were tightly related. Given the nature of the work when I'm doing these kind of changes, I sometimes have to experiment with multiple alternative ways of accomplishing the result I'm looking for looking for the most minimally invasive way of doing so. The above separation of commits is ahistorical and somewhat artificial and the ask requires more additional work from me than what would otherwise be required from the reviewer.

In more general terms, I understand that we have different perspectives on where different changes fall in the "increased complexity"/"user focus" axes, but I must say, I do not appreciate the antagonistic way you interact with me, here or in the past. If you desire to talk about it, with mods or on our own, you can find me in Zulip. I do not believe this is the appropriate place for that discussion.

petrochenkov commented:
@estebank 
I prefer transparency and no mod involvement is required here.
I've been talking about this not once, not twice, not even three times, but yesterday evening I opened this PR and saw all the same stuff, and it was pretty infuriating.
I'll review the new version in a couple of days.

fee1-dead commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -1634,6 +1642,33 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            );
+            // If we suggest importing a public re-export, don't point at the definition.
+            if point_to_def && ident.span != outer_ident.span {
+                err.span_label(
+                    outer_ident.span,
+                    format!("{} `{}` is not publicly reachable", this_res.descr(), ident),
+                );
```
can this use translatable diagnostics?

estebank replied:
I'll do this as part of a separate PR.

fee1-dead commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -2724,6 +2724,32 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
+                        if let DefKind::Trait = tcx.def_kind(item_def_id) && !visible_item {
+                            // FIXME(estebank): extend this to search for all the types that do
+                            // implement this trait and list them.
+                            err.note(format!(
+                                "`{short_item_name}` is a \"sealed trait\", because to implement \
+                                 it you also need to implelement `{}`, which is not accessible; \
+                                 this is usually done to force you to use one of the provided \
+                                 types that already implement it",
+                                with_no_trimmed_paths!(tcx.def_path_str(def_id)),
+                            ));
```
same here, can you use translatable diagnostics instead?

petrochenkov replied:
https://github.com/rust-lang/rust/pull/112686#discussion_r1237831728

petrochenkov commented:
Some more detailed meta-comments to clarify what I'd like to see from PRs like this as a reviewer:
- 9f1279030655143b244aeb0f750641654a8da969 34c7c130987c12f11db222743da8bf62ed352e42 761d2b8c5937d72d3622e4664bf3280875e5df6d
  - These commits both add new tests and change their output, so it's hard to see what changes.
  - From a PR author's point of view it would be easiest to just dump all the new tests into the single first commit as I suggested in https://github.com/rust-lang/rust/pull/112686#issuecomment-1594990915, and from a reviewer's point of view that would be totally fine.
- 55ef88eb8ab59351eb9a5406c8a51ad66776ae84
  - Nit: moving code (sort) and changing code (dedup) in a same commit, so it's hard to see what changes.
- 9f1279030655143b244aeb0f750641654a8da969 24dbcab241c1e9c1558a840da2ceeaddf27d5aaa 196fbc6d425665072034c4e16cf0e6cf024ec087
  - This change is more independent from the rest of the PR than others, and touches a different area of the compiler maintained by different people, so it would be better submitted as a separate PR.
  - That's probably too much splitting into commits, the commits don't touch non-diagnostic infra and all of them add a single diagnostic in a single place, could be merged into one.
- Otherwise the current split into commits is great and makes it much easier to understand what happens, compared to the original version.
  - It still took me hours to review and write this post though, because there are many "logical units" of changes.
  - Some GitHub review tools work great at PR level, but not at individual commit levels, that's why I ask people to split large PRs (not necessarily large in terms of code amount, but more in terms of these logical units of changes). This PR would be fine two PRs, as I mentioned a few lines above.

I'll post the comments on code a bit later.

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -1634,6 +1642,33 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+        if let Some((this_res, outer_ident)) = outermost_res {
+            let import_suggestions = self.lookup_import_candidates(
+                outer_ident,
+                Namespace::TypeNS,
```
```suggestion
                this_res.ns(),
```

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -2474,19 +2509,19 @@ fn show_candidates(
     path: Vec<Segment>,
     append: &str,
-) {
+) -> bool {
```
Could you add a comment telling what this function returns?

estebank replied:
Added to the docstring

petrochenkov commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -1369,6 +1375,22 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+        // CAUTION! If you add a new return path to this function, you need to ensure that you
+        // recover these into it again, otherwise rustc will start allowing use of items that
+        // should remain unreachable.
+        let mut privacy_errors = std::mem::take(&mut self.privacy_errors);
```
I think something like
```rust
let num_privacy_errors = self.privacy_errors.len();
// ...
for error in &mut self.privacy_errors[num_privacy_errors..] {
    error.outermost_res = last_res;
}
```
would help to avoid all this moving in/out both here and in `imports.rs`.

estebank replied:
I considered that, but I was concerned when I saw some logic that modified `privacy_errors`. Doing it that way provides no recourse to any change to the ordering of `privacy_errors` or removal of elements that invalidate the range. If this not happening is something that we feel comfortable relying on, maybe only with warnings on the field definition itself, then doing it that way is valid.

petrochenkov replied:
I think its fine, `privacy_erros` are supposed to be append-only.

The only `privacy_errors.truncate(..)` is a recent addition (https://github.com/rust-lang/rust/pull/109788) for another error deduplication, maybe it's better to replace it with some span-based deduplication too to keep the append-only property (not in this PR).
It's "local" in the sense that we are popping what we just pushed, so it's not visible from the outside of `resolve_qpath`.

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -1634,6 +1642,33 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+                "",
+            );
+            // If we suggest importing a public re-export, don't point at the definition.
+            if point_to_def && ident.span != outer_ident.span {
```
```suggestion
            if point_to_def {
```
Not clear what the second condition came from, what case does it cover?

estebank replied:
Avoids

```
LL | impl a::B for S {}
   |         ^
   |         |
   |         private trait
   |         trait `B` is not publicly reachable
```

The effects aren't seen because it is the lack of changes to the existing tests (and when splitting the commits this condition wasn't something that I mentally flagged as noteworthy when cherrypicking changes).

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -1634,6 +1642,33 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            if point_to_def && ident.span != outer_ident.span {
+                err.span_label(
+                    outer_ident.span,
+                    format!("{} `{}` is not publicly reachable", this_res.descr(), ident),
```
```suggestion
                    format!("{} `{}` is not publicly reexported", this_res.descr(), ident),
```
This is the usual wording for this, "reachable" is already used for something else in the privacy area.

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -2500,20 +2535,24 @@ fn show_candidates(
     if !accessible_path_strings.is_empty() {
-        let (determiner, kind, name) = if accessible_path_strings.len() == 1 {
-            ("this", accessible_path_strings[0].1, format!(" `{}`", accessible_path_strings[0].0))
+        let (determiner, kind, name, post) = if accessible_path_strings.len() == 1 {
```
Maybe can use
```suggestion
        let (determiner, kind, name, post) = if let [first_element] = accessible_path_strings {
```
form here, should look better.

petrochenkov commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -926,11 +934,34 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+                &path,
+                None,
+                &import.parent_scope,
+                Some(Finalize { report_private: ReportPrivate::Skip, ..finalize }),
```
Not clear why introducing `ReportPrivate::Skip` is necessary.
Aren't these errors already deduplicated due to `dedup_span`?

estebank replied:
Calling `resolve_path` with `ReportPrivate::Error` will bail early on resolution errors and not return the `Res`/`DefId` we need. Calling it with `ReportPrivate::Report` will give us back the `Res` but will also produce spurious errors.

petrochenkov replied:
This is speculative resolution, just to get the resolution result, right?
In that case it doesn't need `Finalize` at all, and `None` can be used here.

estebank replied:
Using `None` [worsens the case for sealed traits in foreign crates](https://github.com/rust-lang/rust/pull/112686/commits/b1fd6ac37810a95a4de6af63cbbb4f09f4d355c0#diff-3d63921e1b98fed9fd7bc43bdf56adb40750b22cfbbb9b7919c5a7b5d5197e82). If you prefer we can land the other changes, but reduces the benefit of that part of this change significantly, since the people who would need that note would be when encountering it in dependencies only. Maybe there's an alternative targeted tweak in `resolve_path` that we can keep the desired behavior for foreign crates too?

petrochenkov replied:
Looks like `None` always behaves like `ReportPrivate::Error` here which is not what we need.
(I don't think it correlates with items being from foreign crates though.)

petrochenkov replied:
I'll think what to do with this.

estebank replied:
Yeah, you're right it's not because of foreign crates, it's because of the behavior around imports. The incorrect `if !self.privacy_errors.is_empty()` was making this new logic not trigger anymore. It now points at the module immediately after, but that's something I can look into after rebasing on top of your latest changes.

estebank commented on tests/ui/privacy/privacy1.rs:
```diff
@@ -129,9 +129,7 @@ mod foo {
-        use bar::baz::{foo, bar};
-        //~^ ERROR: module `baz` is private
-        //~| ERROR: module `baz` is private
+        use bar::baz::{foo, bar}; //~ ERROR: module `baz` is private
```
I'm realizing that outside of the test suite this was likely previously caught by the generic deduplication logic. I'd noticed the duplicated line in the test and it seemed reasonable to avoid it, but this likely never had user impact.

petrochenkov replied:
Could you remove this new deduplication?

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -2500,20 +2536,24 @@ fn show_candidates(
     if !accessible_path_strings.is_empty() {
-        let (determiner, kind, name) = if accessible_path_strings.len() == 1 {
-            ("this", accessible_path_strings[0].1, format!(" `{}`", accessible_path_strings[0].0))
+        let (determiner, kind, name, post) = if let [first_element] = &accessible_path_strings[..] {
```
```suggestion
        let (determiner, kind, name, through) = if let [(name, descr, _, _, via_import)] = &accessible_path_strings[..] {
```

petrochenkov commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -1522,6 +1531,7 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                     let maybe_assoc = opt_ns != Some(MacroNS) && PathSource::Type.is_expected(res);
```
I think you can move the whole
```rust
        for error in &mut self.privacy_errors[privacy_errors_len..] {
            error.outermost_res = Some((res, ident));
        }
```
loop to this point (right after `let res = ...`).

estebank replied:
Wouldn't this mean that we're doing `O(privacy_errors_len * path_segments_len)` operations? The first condition, `if let Some(net_module) = binding.module()` will fallthrough and move to the next path segment, iterating over the privacy errors and setting them again (until we return).

petrochenkov replied:
`self.privacy_errors[privacy_errors_len..].len()` is zero during successful compilation, so we don't slow down the good path.

And even on a bad path, where perf doesn't matter much, there should be about `1 * path_segments_len` operations, which is nothing.

petrochenkov commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -931,6 +935,23 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            if let PathResult::Module(ModuleOrUniformRoot::Module(module)) =
+                self.resolve_path(&path, None, &import.parent_scope, None, ignore_binding)
+            {
+                res = module.res();
```
The loop below can be moved to this point.
```rust
        for error in &mut self.privacy_errors[privacy_errors_len..] {
            error.outermost_res = module.res().map(|res| (res, ident));
        }
```

petrochenkov commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -931,6 +935,23 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             };
 
+        let mut res = None;
+        if !self.privacy_errors.is_empty() {
```
```suggestion
        if self.privacy_errors.len() != privacy_errors_len {
```

petrochenkov commented on compiler/rustc_resolve/src/ident.rs:
```diff
@@ -1568,7 +1578,6 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
                             ));
                         }
                     }
-
```
Nit: spurious change.

petrochenkov commented on compiler/rustc_resolve/src/imports.rs:
```diff
@@ -931,6 +935,23 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+            let mut path = import.module_path.clone();
+            path.push(Segment::from_ident(ident));
+            if let PathResult::Module(ModuleOrUniformRoot::Module(module)) =
+                self.resolve_path(&path, None, &import.parent_scope, None, ignore_binding)
```
```suggestion
                self.resolve_path(&path, None, &import.parent_scope, Some(finalize), ignore_binding)
```
Btw, what happens if you use the same `finalize` as above (with `ReportPrivate::Report`).
The dedup span will be the same as in resolution above (`import.span`), so the reported errors should be deduplicated into nothing?

estebank replied:
This actually solves the issue I was seeing in https://github.com/rust-lang/rust/pull/112686#discussion_r1237360321. Changed.

estebank commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -136,9 +137,14 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
-            if reported_spans.insert(error.dedup_span) {
-                self.report_privacy_error(error);
+        self.privacy_errors.sort_by_key(|e| e.ident.span);
+        for error in std::mem::take(&mut self.privacy_errors) {
```
Because `fn report_privacy_error(&mut`, which is needed to be able to call `lookup_import_candidates` from within it, iterating over `privacy_errors` without taking ownership of them is not possible. I thought that `privacy_errors` was not being used for anything after the errors were reported. Was this understanding wrong?

petrochenkov replied:
Ah, okay, the signature of `report_privacy_error` was changed to `&mut self`, I didn't notice.
I though the change ended up being unnecessary.
(`privacy_errors` can be taken out at this point, yes.)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5337252253/jobs/9673038824) [(plain)](https://github.com/rust-lang/rust/commit/e2bcbbcba9f128604f84d9653e4d27f9acf9e56c/checks/14446977072/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0599]: no associated item named `Trait` found for struct `rustc_middle::ty::Clause` in the current scope
    --> compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:2728:44
     |
2728 |                         && let ty::Clause::Trait(trait_pred) = clause
     |                                            ^^^^^ associated item not found in `Clause<'_>`
For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustc_trait_selection` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:00:50
```

</details>

petrochenkov commented on compiler/rustc_resolve/src/lib.rs:
```diff
@@ -2036,6 +2039,12 @@ fn module_to_string(module: Module<'_>) -> Option<String> {
 }
 
+#[derive(Clone, Copy, Debug)]
+enum ReportPrivate {
```
Introducing this enum is no longer necessary.

petrochenkov commented on compiler/rustc_resolve/src/diagnostics.rs:
```diff
@@ -2493,20 +2529,25 @@ fn show_candidates(
-            ("one of these", "items", String::new())
-        };
+        let (determiner, kind, name, through) =
+            if let [first_element] = &accessible_path_strings[..] {
```
```suggestion
            if let [(name, descr, _, _, via_import)] = &accessible_path_strings[..] {
```

petrochenkov commented:
r=me after addressing https://github.com/rust-lang/rust/pull/112686#discussion_r1238627309 and cleaning up git history.
@rustbot author

estebank commented:
Addressed the last review comments. Rebased on top of a recent master to incorporate #112892.

Do you want me to merge everything into a single commit?

petrochenkov commented:
Well, ideally it would still be commit per independent change, but the PR is now much smaller so it's not too bad as is.
@bors r+

bors commented:
:pushpin: Commit 7dffd24da5292cc01f18c4ab53f0b96f682fcbe3 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 7dffd24da5292cc01f18c4ab53f0b96f682fcbe3 -->
<!-- homu: {"type":"Approved","sha":"7dffd24da5292cc01f18c4ab53f0b96f682fcbe3","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7dffd24da5292cc01f18c4ab53f0b96f682fcbe3 with merge 04075b32021932e3e8f6ab55d519b3b3494b6ef9...
<!-- homu: {"type":"BuildStarted","head_sha":"7dffd24da5292cc01f18c4ab53f0b96f682fcbe3","merge_sha":"04075b32021932e3e8f6ab55d519b3b3494b6ef9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5349016190/jobs/9703121214)
Approved by: petrochenkov
Pushing 04075b32021932e3e8f6ab55d519b3b3494b6ef9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5349016190/jobs/9703121214"},"merge_sha":"04075b32021932e3e8f6ab55d519b3b3494b6ef9"} -->

rust-timer commented:
Finished benchmarking commit (04075b32021932e3e8f6ab55d519b3b3494b6ef9): [comparison URL](https://perf.rust-lang.org/compare.html?start=2efe09170530fa18e42ff05b8d9dd23f00b5c430&end=04075b32021932e3e8f6ab55d519b3b3494b6ef9&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2efe09170530fa18e42ff05b8d9dd23f00b5c430&end=04075b32021932e3e8f6ab55d519b3b3494b6ef9&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2efe09170530fa18e42ff05b8d9dd23f00b5c430&end=04075b32021932e3e8f6ab55d519b3b3494b6ef9&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2efe09170530fa18e42ff05b8d9dd23f00b5c430&end=04075b32021932e3e8f6ab55d519b3b3494b6ef9&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.4% | [1.0%, 1.8%] | 3     |
| Regressions ❌ <br /> (secondary)  | 3.1% | [2.5%, 3.7%] | 6     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.4% | [1.0%, 1.8%] | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=2efe09170530fa18e42ff05b8d9dd23f00b5c430&end=04075b32021932e3e8f6ab55d519b3b3494b6ef9&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 661.969s -> 663.045s (0.16%)
